api/krusty/component_test.go

```diff
@@ -0,0 +1,441 @@
+	writeComponentBase(th)
+	writeComponentPatch(th)
+	writeComponentProd(th)
+	m := th.Run("/app/prod", th.MakeDefaultOptions())
+	th.AssertActualEqualsExpected(m, `
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: patched-storefront
+spec:
+  replicas: 3
+---
+apiVersion: v1
+data:
+  otherValue: "10"
+  patchValue: "5"
+  testValue: "2"
+kind: ConfigMap
+metadata:
+  annotations: {}
+  labels: {}
+  name: patched-my-configmap-5g55h28cdh
+---
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: patched-db
+spec:
+  type: Logical
+---
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: patched-stub
+spec:
+  replicas: 1
+`)
+}
+
+func TestMultipleComponents(t *testing.T) {
+	th := kusttest_test.MakeHarness(t)
+	writeComponentBase(th)
+	writeComponentPatch(th)
+	writeComponentProd(th)
+	th.WriteF("/app/additionalpatch/kustomization.yaml", `
+apiVersion: kustomize.config.k8s.io/v1alpha1
+kind: Component
+configMapGenerator:
+- name: my-configmap
+  behavior: merge
+  literals:	
+    - otherValue=9
+`)
+	th.WriteK("/app/prod", `
+resources:
+- ../base
+- db
+
+components:
+- ../patch
+- ../additionalpatch
+`)
+	m := th.Run("/app/prod", th.MakeDefaultOptions())
+	th.AssertActualEqualsExpected(m, `
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: patched-storefront
+spec:
+  replicas: 3
+---
+apiVersion: v1
+data:
+  otherValue: "9"
+  patchValue: "5"
+  testValue: "2"
+kind: ConfigMap
+metadata:
+  annotations: {}
+  labels: {}
+  name: patched-my-configmap-9fddc87cdk
+---
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: patched-db
+spec:
+  type: Logical
+---
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: patched-stub
+spec:
+  replicas: 1
+`)
+}
+
+func TestNestedComponents(t *testing.T) {
+	th := kusttest_test.MakeHarness(t)
+	writeComponentBase(th)
+	writeComponentPatch(th)
+	writeComponentProd(th)
+	th.WriteF("/app/additionalpatch/kustomization.yaml", `
+apiVersion: kustomize.config.k8s.io/v1alpha1
+kind: Component
+components:
+- ../patch
+configMapGenerator:
+- name: my-configmap
+  behavior: merge
+  literals:	
+    - otherValue=9
+`)
+	th.WriteK("/app/prod", `
+resources:
+- ../base
+- db
+
+components:
+- ../additionalpatch
+`)
+	m := th.Run("/app/prod", th.MakeDefaultOptions())
+	th.AssertActualEqualsExpected(m, `
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: patched-storefront
+spec:
+  replicas: 3
+---
+apiVersion: v1
+data:
+  otherValue: "9"
+  patchValue: "5"
+  testValue: "2"
+kind: ConfigMap
+metadata:
+  annotations: {}
+  labels: {}
+  name: patched-my-configmap-9fddc87cdk
+---
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: patched-db
+spec:
+  type: Logical
+---
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: patched-stub
+spec:
+  replicas: 1
+`)
+}
+
+// If a patch sets a name prefix on a base, then that base can also be separately included
+// without being affected by the patch in another branch of the resource tree
+func TestBasicComponentWithRepeatedBase(t *testing.T) {
```

Comment:
This is a great test.

Would you also add tests for:

- 2 components add the same base / resource
  - kustomization has 2 components
  - each component adds the same resource
  - this should either 1) error or 2) de-dup the resource (docs should include this)
- 2 bases add the same component
  - kustimization has 2 bases
  - each base adds the same components
  - this should work

Reply:
Thanks.  I've added some more tests: 

> 2 components add the same base / resource
> kustomization has 2 components
> each component adds the same resource
> this should either 1) error or 2) de-dup the resource (docs should include this)

`components-can-add-the-same-base-if-the-first-renames-resources`

These will error (unless the first component explicitly de-dups the resource) - see below.


> 2 bases add the same component
> * kustimization has 2 bases
> * each base adds the same components
> * this should work

`components-cannot-add-the-same-resource`, `components-cannot-add-the-same-base`, `components-cannot-add-bases-containing-the-same-resource`, `multiple-bases-can-add-the-same-component-if-it-doesn-not-define-named-entities`

If the components are not defining resources (just modifying configmaps, etc.) then this would work.  However, if a component defines a deployment (for example) and is included by multiple bases of a kustomization, then the build will fail because we will get duplicate deployments, and the bases would need to rename the deployment to disambiguate them.  This happens because when Kustomize processes a Kustomization or Component, it loads and processes that file anew - it does not look to see if it has already processed it another part of the Kustomization tree.  It is up to the user to modify any duplicate resources, e.g. by changing their names.  Is that not what you expected?  I think some graph diagrams to show how the Kustomizations and Components interact might help?